File: src/config/predicates.ts
import data from '../../shared/predicates.json';

export interface PredicateSpec {
    name: string;
    description: string;
    example: string;
    asymmetric: boolean;
    dag: boolean;
}

const SPECS: PredicateSpec[] = data.predicates;

export const PREDICATE_SPECS = SPECS;
export const PREDICATE_NAMES = SPECS.map((p) => p.name);
export const PREDICATE_LIST_PIPE = PREDICATE_NAMES.join(' | ');
export const ALLOWED_PREDICATES = new Set<string>(PREDICATE_NAMES);
export const ASYMMETRIC_PREDICATES = new Set<string>(
    SPECS.filter((p) => p.asymmetric).map((p) => p.name)
);
export const DAG_RELATIONS = new Set<string>(
    SPECS.filter((p) => p.dag).map((p) => p.name)
);

/** Human-readable rubric block, drop-in for the KG_EXTRACTOR prompt. */
export const PREDICATE_RUBRIC = SPECS.map((p) => {
    const head = `- ${p.name.padEnd(18)}→ ${p.description}`;
    return p.example ? `${head} ("${p.example}")` : head;
}).join('\n');

export type AllowedRelation = typeof PREDICATE_NAMES[number];

File: backend/graph/predicates.py
"""
Single source of truth for the predicate ontology.

Loads from /shared/predicates.json so that the TS kg-builder and the
Python concept_extractor cannot drift out of sync.
"""
import json
from dataclasses import dataclass
from pathlib import Path

_SHARED = Path(__file__).resolve().parents[2] / "shared" / "predicates.json"


@dataclass(frozen=True)
class PredicateSpec:
    name: str
    description: str
    example: str
    asymmetric: bool
    dag: bool


def _load() -> list[PredicateSpec]:
    with _SHARED.open() as f:
        data = json.load(f)
    return [PredicateSpec(**item) for item in data["predicates"]]


PREDICATE_SPECS: list[PredicateSpec] = _load()
PREDICATE_NAMES: list[str] = [p.name for p in PREDICATE_SPECS]
ALLOWED_PREDICATES: set[str] = set(PREDICATE_NAMES)
ASYMMETRIC_PREDICATES: set[str] = {p.name for p in PREDICATE_SPECS if p.asymmetric}
DAG_RELATIONS: set[str] = {p.name for p in PREDICATE_SPECS if p.dag}


def predicate_rubric() -> str:
    """Human-readable rubric for inclusion in extractor prompts."""
    lines = []
    for p in PREDICATE_SPECS:
        line = f"- {p.name:18s}→ {p.description}"
        if p.example:
            line += f' ("{p.example}")'
        lines.append(line)
    return "\n".join(lines)